test/jdk/tools/jpackage/helpers/jdk/jpackage/test/TestMethodSupplier.java

```diff
@@ -0,0 +1,512 @@
+            final Class methodClass = Class.forName(className);
+
+            // Get the list of all public methods as need to deal with overloads.
+            return Stream.of(methodClass.getMethods()).filter(method -> {
+                return method.getName().equals(methodName);
+            }).toList();
+        }
+
+        static MethodQuery fromQualifiedMethodName(String qualifiedMethodName) {
+            int lastDotIdx = qualifiedMethodName.lastIndexOf('.');
+            if (lastDotIdx == -1) {
+                throw new IllegalArgumentException("Class name not specified");
+            }
+
+            var className = qualifiedMethodName.substring(0, lastDotIdx);
+            var methodName = qualifiedMethodName.substring(lastDotIdx + 1);
+
+            return new MethodQuery(className, methodName);
+        }
+    }
+
+    List<Method> findNullaryLikeMethods(MethodQuery query) throws NoSuchMethodException {
+        List<Method> methods;
+
+        try {
+            methods = query.lookup();
+        } catch (ClassNotFoundException ex) {
+            throw new NoSuchMethodException(
+                    String.format("Class [%s] not found", query.className()));
+        }
+
+        if (methods.isEmpty()) {
+            throw new NoSuchMethodException(String.format(
+                    "Public method [%s] not found in [%s] class",
+                    query.methodName(), query.className()));
+        }
+
+        methods = methods.stream().filter(method -> {
+            if (isParameterized(method) && isTest(method)) {
+                // Always accept test method with annotations producing arguments for its invocation.
+                return true;
+            } else {
+                return method.getParameterCount() == 0;
+            }
+        }).filter(this::isEnabled).toList();
+
+        if (methods.isEmpty()) {
+            throw new NoSuchMethodException(String.format(
+                    "Suitable public method [%s] not found in [%s] class",
+                    query.methodName(), query.className()));
+        }
+
+        return methods;
+    }
+
+    boolean isTestClass(Class<?> type) {
+        var typeStatus = processedTypes.get(type);
+        if (typeStatus == null) {
+            typeStatus = Verifier.isTestClass(type) ? TypeStatus.TEST_CLASS : TypeStatus.NOT_TEST_CLASS;
+            processedTypes.put(type, typeStatus);
+        }
+
+        return !TypeStatus.NOT_TEST_CLASS.equals(typeStatus);
+    }
+
+    void verifyTestClass(Class<?> type) throws InvalidAnnotationException {
+        var typeStatus = processedTypes.get(type);
+        if (typeStatus == null) {
+            try {
+                Verifier.verifyTestClass(type);
+                processedTypes.put(type, TypeStatus.VALID_TEST_CLASS);
+                return;
+            } catch (InvalidAnnotationException ex) {
+                processedTypes.put(type, TypeStatus.TEST_CLASS);
+                throw ex;
+            }
+        }
```

Comment:
Maybe move this code into `switch` statement under `default`? Also, should we use `TypeStatus.UNKNOWN` instead of `null`?

Reply:
> Maybe move this code into switch statement under default?

Can't use "null" in the switch.

> should we use TypeStatus.UNKNOWN instead of null?

To deal with `TypeStatus.UNKNOWN` and `null`? Besides of that what is the semantics of `TypeStatus.UNKNOWN`? 

- `TypeStatus.NOT_TEST_CLASS` is for types that are NOT test classes.
- `TypeStatus.VALID_TEST_CLASS` is for types that are valid test classes.
- `TypeStatus.TEST_CLASS` is for all other types, i.e. invalid test classes.

How does `TypeStatus.UNKNOWN` fit?

Reply:
Never mind. I see why it is `null`.